operator: Add LB-IPAM

dylandreimerink opened the pull request:
This PR introduces a new feature called LoadBalancer IP Address Management, LB-IPAM for short. The goal of LB-IPAM is to provide a mechanism to allocate and assign IP addresses to services of type `LoadBalancer`, to be used when no external IPAM is available. The primary use-case we are targeting is a private cloud/bare metal deployment where one or more CIDRs are assigned to the cluster to be used. Additional Cilium features should be utilized to direct traffic for the assigned IPs to the cluster and the correct nodes(for now just BGP, but not exclusively).

LB-IPAM has the concept of an IP Pool which is a list of CIDRs and an optional service label selector. Users can add, modify or remove these at will via the `CiliumLoadBalancerIPPool` custom resource (formally the `CiliumBGPLoadBalancerIPPool`, which was renamed to make it more generic). The service label selector can be used to limit which services can get IPs from the given pool.

LB-IPAM can allocate and assign both IPv4 and IPv6 addresses in SingleStack and DualStack operation. LB-IPAM can be deployed in a cluster with multiple load balancers since it respects the loadBalancerClass values of services. Lastly LB-IPAM allows services to request one or more specific IPs. 

```release-note
Add LoadBalancer IP address management (LB-IPAM)
```

dylandreimerink committed:
k8s: Rename and reuse BGP IP Pool

This commit renames the CiliumBGPLoadBalancerIPPool CRD to the
CiliumLoadBalancerIPPool so it may be used for load balancers other
than those who use BGP.

The IP Pool will be used by the operators LB IPAM component, and the
contents of the CRD have been updated to match the new requirements.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1730 @@
+}
+
+func (cfg LBIPAMConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("lb-ipam", cfg.EnableLBIPAM, "Enables LB IPAM, which manages IPs for LoadBalancer services internally without cloud providers")
```
Why make this a flag? If no CRs are present, there's nothing to do...

dylandreimerink replied:
LB-IPAM assumes it is the default service LB IP provider, in case we deploy in k8s <v1.24. Thus, if LB-IPAM is enabled, it will start modifying service statuses even if there are no pools(marking them with conditions explaining that we can't allocate IPs since there are no pools). Hence why it is behind a feature flag

squeed replied:
I would rather the controller do nothing if there are no pools created.

You can even be clever and not start up the Service informer until the first pool is created, if you're concerned about resource usage.

dylandreimerink replied:
Alright, made it so LB-IPAM wakes up when it sees a pool and will start doing its work and goes back to sleep when the last pool has been removed.

squeed commented on pkg/k8s/apis/cilium.io/v2alpha1/lbipam_types.go:
```diff
@@ -0,0 +1,126 @@
+// +deepequal-gen=false
+
+// CiliumLoadBalancerIPPoolStatus contains the status of a CiliumLoadBalancerIPPool.
+type CiliumLoadBalancerIPPoolStatus struct {
```
In general, a *Status struct should have an [array of conditions](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3883). Conficting could be one of the conditions.

dylandreimerink replied:
I did initially have conditions, but switched to this since it allows me to display the value in the kubectl output of `kubectl get ippools`. I can add a conditions array for compatibility reasons.

squeed replied:
If possible, please only add a conditions array, then configure kubectl out via the AdditionalPrinterColumns field. Inflexible status fields like these always come back to bite us.

dylandreimerink replied:
Alright, got everything to work with conditions, how does this look?

rolinh reviewed (approved):
Vendor changes lgtm

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+however, when deploying in a private cloud environment, these facilities are not
+always available.
+
+LB IPAM work in conjunction with features like the :ref:`bgp_control_plane`. Where
```
```suggestion
LB IPAM works in conjunction with features like the :ref:`bgp_control_plane`. Where
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+other features are responsible for load balancing and/or advertisement of these
+IPs. 
+
+LB IPAM is implicitly enabled when dependant features are enabled.
```
Which are they?

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+LB IPAM has the notion of IP Pools which the administrator can create to tell 
+Cilium which IP ranges can be used to allocate IPs from.
+
+A basic IP Pools with both an IPv4 and IPv6 range looks like.
```
```suggestion
A basic IP Pools with both an IPv4 and IPv6 range looks like this:
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+    blue-pool   false      false         65788           2s
+
+.. note::
+    The amount of available IPs in the pool is less than the actual sum of all
```
```suggestion
    The amount of available IPs in the pool is lower than the actual sum of all
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+.. note::
+    The amount of available IPs in the pool is less than the actual sum of all
+    usable IPs in the CIDRs because the allocation logic is limited to 65536 IPs
+    per CIDR. CIDRs containing more than 65536 IPs can be broken up into multiple
```
Not sure, but I'd rather say “broken down”?

```suggestion
    per CIDR. CIDRs containing more than 65536 IPs can be broken down into multiple
```

dylandreimerink replied:
I don't know honestly. If I say "breaking up a problem" I mean as in smaller pieces just like "breaking up a group". but "breaking down a problem" is also solving in smaller parts, like breaking "down a building". Lets go with "down" for now.

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+io.kubernetes.service.name      ``.meta.name``
+=============================== ===================
+
+.. code-block:: yaml
```
```suggestion
For example:

.. code-block:: yaml
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+IP Pools are not allowed to have overlapping CIDRs. When an administrator does
+create pools which overlap, a soft error is caused. The last added pool will be
+marked as ``Conflicting`` and no further allocation will happen from that pool.
+Therefor administrators should always check the status of all pools after making
```
```suggestion
Therefore, administrators should always check the status of all pools after making
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+Therefor administrators should always check the status of all pools after making
+modifications.
+
+If we for example add 2 pools(``blue-pool`` and ``red-pool``) both with the same
```
I won't let go regarding parenthesis :)

```suggestion
For example, if we add 2 pools (``blue-pool`` and ``red-pool``) both with the same
```

dylandreimerink replied:
Sigh. You shouldn't. Just just to writing `if(`, `for(` and `somefunc(`. Doesn't translate that will into non-code formatting.

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+------
+
+The IP Pool's status contains additional counts which can be used to monitor
+the amount of used and available IPs. A machine parsable output can be gotten like so.
```
I'm not familiar with “gotten”, but that's maybe OK. I'm nitpicking here.

```suggestion
the amount of used and available IPs. A machine parsable output can be obtained like so.
```

dylandreimerink replied:
`gotten` is more American English(I tend to use both since it is not my primary language), `obtained` does sound better, will go with that.

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+
+Kubernetes >= v1.24 supports `multiple load balancers <https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`_ 
+in the same cluster. Picking between load balancers is done with the ``.spec.loadBalancerClass`` field. 
+When LB IPAM is enabled it assumed that it is supposed to allocate and assign IPs for services with 
```
```suggestion
When LB IPAM is enabled it allocates and assigns IPs for services with 
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+When LB IPAM is enabled it assumed that it is supposed to allocate and assign IPs for services with 
+no load balancer class set.
+
+LB IPAM only does IP allocation and doesn't provide load balancing services by itself. Therefor
```
```suggestion
LB IPAM only does IP allocation and doesn't provide load balancing services by itself. Therefore,
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+Requesting IPs
+--------------
+
+Services can request specific IPs. The legacy wait of doing so is via ``.spec.loadBalancerIP``
```
```suggestion
Services can request specific IPs. The legacy way of doing so is via ``.spec.loadBalancerIP``
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+until its future removal.
+
+The new way of requesting specific IPs is to use annotations, ``io.cilium/lb-ipam-ips`` in the case
+of Cilium LB IPAM. This annotation takes a comma separated list of IP addresses, allowing for
```
```suggestion
of Cilium LB IPAM. This annotation takes a comma-separated list of IP addresses, allowing for
```

qmonnet commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,382 @@
+multiple IPs to be requested at once.
+
+The service selector of the IP Pool still applies, requested IPs will not be allocated or assigned
+if the services doesn't match the pool's selector.
```
```suggestion
if the services don't match the pool's selector.
```

qmonnet reviewed (changes requested):
Code looks OK, I've spotted some minor things for the docs.

Renaming the CRD that `CiliumBGPLoadBalancerIPPoolList` can no longer be used, right? Is that OK to simply remove this without deprecating first? And should it be mentioned in `Documentation/operations/upgrade.rst`?

qmonnet reviewed (approved):
Thanks!

tommyp1ckles commented on install/kubernetes/cilium/Chart.yaml:
```diff
@@ -118,10 +118,9 @@ annotations:
       version: v2alpha1
-      name: ciliumbgploadbalancerippools.cilium.io
-      displayName: Cilium BGP Load Balancer IP Pool
+      name: ciliumloadbalancerippools.cilium.io
```
Renaming a CRD seems like a potentially breaking change. Is there going to be an upgrade path for this?

dylandreimerink replied:
The current CRD was never put into service. No code ever use it and it wasn't documented as having any utility. I wouldn't expect users to have CRs in their cluster. If they do, and we update, there isn't any consequence.

aanm commented on install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:
```diff
@@ -82,9 +82,11 @@ rules:
   - list
   - watch
+{{- if or .Values.ingressController.enabled (and .Values.bgpControlPlane.enabled .Values.bgp.announce.loadbalancerIP) }} 
+  - update
```
Jus to be sure, do we need to update endpoints as well? Why?

dylandreimerink replied:
No, that was an oversight. I discovered that we have a rule which grants 'update' permissions to `services/status` for the MetalLB. Turns out it didn't work since I patch the status. This change also wouldn't have worked after the code switched from update to patch. 

I have updated the LB rule to also include patching.

squeed replied:
I'd suggest just always allowing patch + update, and drop the Helm conditional. It simplifies enabling this in day-2 scenarios.

(In fact, I'm arguing for removing the enable flag all together, but that's a separate thread :-) )

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
```
This will stop processing as soon as one of the channels is closed; did you mean `poolChan != nil || svcChan != nil` ?

dylandreimerink replied:
No, the resources expect both channels to be drained on shutdown, this ensures we keep reading until we get a close signal from both channels. I will reword the comment to make this clearer

squeed replied:
Right; I think you have the logic backwards. It should be a `||` and not an `&&`.

joamaki replied:
I think @squeed is right. We need `poolChan != nil || svcChan != nil`, e.g. keep going until both are non-nil.

dylandreimerink replied:
Ah, I see it now. Sorry about that one

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+		case event := <-poolChan:
+			event.Handle(
+				func() error {
+					err := ipam.settleConflicts()
```
Shouldn't this only be called once after the store is synced?

squeed replied:
Nevermind, that's already how it works :-)

dylandreimerink replied:
The first callback is indeed the 'onSync' callback. The resources work in such a way that we first get an upsert for every entry in the resource and then the onSync.

So for example:
* onUpsert pool 1 (event from List)
* onUpsert pool 2 (event from List)
* onSync (Initial sync done)
* onUpsert pool 2 (event from Watch)

We need to call settle conflicts on any change to the pools, not just on sync, the fact that we don't do it during initialization is just a small optimization.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+				func(k resource.Key, svc *core_v1.Service) error {
+					err := ipam.handleUpsertService(svc)
+					if err != nil {
+						syncErrors = append(syncErrors, err)
```
Why wouldn't you want to re-try here? What if there's a brief blip in apiserver connectivity? Surely we should just retry in that case, right?

joamaki replied:
Yep, I'd also just retry all errors instead (e.g. return non-nil from handler). Seems like we just now return the error, log it and then just stop which may cause a transient connectivity failure to leave the operator in a broken state requiring the admin to manually restart the operator process.

dylandreimerink replied:
Agreed. I now implemented custom retry logic for the pools. This is because I need to be 100% sure that all pools are synced without errors before continuing to process services, and the resource retry mechanism doesn't have a way to give that guarantee. For the services I simply retry using the resources built-in retry mechanism.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+// do not belong on the service and will move the service to the satisfied or unsatisfied service view store depending
+// on if the service requests are satisfied or not.
+func (ipam *LBIPAM) handleUpsertService(svc *core_v1.Service) error {
+	if svc.GetUID() == "" {
```
We shouldn't use UIDs -- this can go.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+func (ipam *LBIPAM) handleNewPool(pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Sanity check that we do not yet know about this pool.
+	if _, found := ipam.poolStore.GetByName(pool.GetName()); found {
+		ipam.logger.Warnf("LB IPPool with uid '%s' has been created, but a LB IP Pool with the same uid already exists", pool.GetUID())
```
nit: log name, not uid here. UIDs are opaque and non-selectable.

joamaki replied:
And use `WithField` and friends.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+		}
+
+		for _, poolInner := range ipam.poolStore.pools {
+			if poolOuter.GetUID() == poolInner.GetUID() {
```
This should check name, not UID.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+
+		poolConflict := false
+		for _, poolInner := range ipam.poolStore.pools {
+			if poolOuter.GetUID() == poolInner.GetUID() {
```
same; check name.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+		extCIDR := extRange.allocRange.CIDR()
+		found := false
+		for _, newCIDR := range newCIDRs {
+			if newCIDR.IP.Equal(extCIDR.IP) && bytes.Equal(newCIDR.Mask, extCIDR.Mask) {
```
suggestion: Add an Equal function on the Range type.

squeed commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+) (
+	newIP *net.IP,
+	chosenRange *LBRange,
+	full bool,
```
Why not just return a specific constant Error for the "no pool" and "all pools full" state?

squeed commented on operator/pkg/lbipam/range_store.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func intersect(n1, n2 net.IPNet) bool {
+	return n2.Contains(n1.IP) || n1.Contains(n2.IP)
```
I've written logic like this before a few times, and for some reason, I never realized it could be this simple. Duh. Nice :-)

squeed commented on operator/pkg/lbipam/range_store.go:
```diff
@@ -0,0 +1,150 @@
+	for _, outerRange := range outerRanges {
+		for _, innerRange := range innerRanges {
+			// IPs of dissimilar IP families can't overlap
+			outerIsIpv4 := outerRange.allocRange.CIDR().IP.To4() != nil
```
Maybe add a convenience Is4() function to `range`? Then the skip is just `outerRange.Is4() != innerRange.Is4()`. Just a cosmetic thing.

joamaki commented on operator/cmd/resources.go:
```diff
@@ -0,0 +1,33 @@
+		func(lc hive.Lifecycle, c k8sClient.Clientset) resource.Resource[*core_v1.Service] {
+			return resource.New[*core_v1.Service](
+				lc,
+				utils.ListerWatcherFromTyped[*core_v1.ServiceList](c.CoreV1().Services("")),
```
With this we'll end up with two caches&informers for services, other one is in `operator/watchers/k8s_service_sync.go`. In difference to here, it's using slim service and it uses the GetServiceListOptionsModifier() to skip "non-headless" services.

1) Can we use slim Service here?
2) Should `GetServiceListOptionsModifier` be used?
3) How difficult would it be to refactor `k8s_service_sync.go` to observe this resource?

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+}
+
+type LBIPAMConfig struct {
+	EnableLBIPAM bool `mapstructure:"lb-ipam"`
```
Why not just name the flag `enable-lb-ipam`? This would also be more in line with existing flags.

dylandreimerink replied:
Removed the flag entirely in favor of starting and stopping the controller when needed

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
```
Why not assign it in `newLBIPAM`? I'd prefer that, or not have `ctx` in `LBIPAM` at all and just pass it down from `Run()`

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+				ipam.svcOnDelete,
+			)
+		case err := <-errChan:
+			ipam.logger.WithError(err).Error("Resource error")
```
We should do more than just log this error. The handlers may choose to retry and we may run out of retries, in which case we must either `Shutdown()`, or restart from clean state, but we should do something and not just leave things hanging.

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+			return err
+		}
+
+		return nil
```
Why are we ignoring "non-retryable" errors? What even is a "non-retriable" error? Why don't we just always retry?

I would really just always retry no matter what goes wrong and allow it to bubble up through the `errChan` in `Run()` and then there decide to either restart or crash, or have infinite retries with exponential back-off. Basically build this so that you never give up,  but rate limit so you don't try too hard. Trying to decide what is and isn't retryable and ignoring errors seems dangerous.

Consider returning a wrapped error with more context and moving the error logging into the retry handler to avoid all the boilerplate that repeats here.

dylandreimerink replied:
All errors now have infinite retries with exponential backoff to a max of 5m, see the operator/cmd/resources.go. That should do it.

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+
+		if isRetryableError(err) {
+			// Return the error, this will trigger the service resource to retry this service at a later time in the hope
```
Comment talks about the "service resource". Perhaps just shorten the comment here to "return the error to retry later"?

joamaki commented on operator/pkg/lbipam/lbipam_fixture_test.go:
```diff
@@ -0,0 +1,353 @@
+	type InitDoneCallback struct {
+		cell.Out
+
+		Callback LBIPAMInitDone `group:"lb-ipam-init-ready"`
```
In general I'd recommend something simpler instead, e.g. like `RegisterOnReady(cb func())` method that would be called from `Invoke` (e.g. before start), or even just `IsReady() bool` and poll that from tests.

joamaki commented on operator/pkg/lbipam/lbipam_fixture_test.go:
```diff
@@ -0,0 +1,353 @@
+
+	// Set the LB-IPAM enable flag.
+	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+	fixture.hive.RegisterFlags(flags)
```
You can also do `hive.Viper().Set("lb-ipam", "true")`

dylandreimerink replied:
Flag has been removed in favor of starting and stopping the controller dynamically

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1728 @@
+			return fmt.Errorf("Error making LB Range for '%s': %w", cidrBlock.Cidr, err)
+		}
+
+		ipam.rangesStore.Add(lbRange)
```
Consider removing rangesStore and just add a "range" index into poolStore? Feels like it'd be easier to have them consistent that way. After all it seems like range is just a property of the pool?

dylandreimerink replied:
Doing so would be a significant refactor and doesn't add a significant amount of clarity at this point, in fact, I would have to make additional types and logic to make it work, so I would prefer to leave the stores as is for the moment.

gandro commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,386 @@
+=============================== ========================
+
+If the ``.spec.loadBalancerClass`` is set to a class which isn't handled by LB IPAM, then LB IPAM
+will ignore the service entirely, not even setting a condition in the status. 
```
nit (to make it a bit more obvious that other components outside of Cilium might handle it):
```suggestion
If the ``.spec.loadBalancerClass`` is set to a class which isn't handled by Cilium's LB IPAM, then Cilium's LB IPAM
will ignore the service entirely, not even setting a condition in the status. 
```

gandro reviewed (commented):
Overall looks very good to me. I'm not very familiar with the new reactive framework yet, so apologize if this is obvious, but I didn't find any code that handles this: What happens if a `patchSvcStatus` or `patchPoolStatus` operations fails (e.g. due to the apiserver temporarily being unavailable)? I was unable to find any retry logic, so will the service or pool never be reconciliated until there is another update event on it?

For pod CIDR ipam, we sync the K8s resource in a controller, ensuring that the new state (e.g. an allocation or status error) is eventually synced/written to K8s. Do you need something like this here as well?

joamaki committed:
k8s/slim: Add missing fields needed by LB-IPAM

This adds:
- metav1.Condition, metav1.ConditionStatus
- metav1.ObjectMeta.Generation
- corev1.IPFamilyPolicy
- corev1.IPFamilyPolicyType
- corev1.LoadBalancerClass
- corev1.Service.{IPFamilyPolicy, LoadBalancerClass}
- corev1.ServiceStatus.Condition

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on operator/cmd/root.go:
```diff
@@ -95,6 +96,14 @@ var (
 			registerOperatorHooks,
 		),
 
+		resourcesCell,
```
`resourcesCell` and `lbipam.Cell` should probably be within `WithLeaderLifecycle`?

joamaki committed:
k8s/resource: Expose the underlying cache.Store in Store[T]

To make it easier to partially transition to using Resource[T],
expose the underlying cache.Store. Hopefully temporary :fingerscrossed:.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1796 @@
+	LC         hive.Lifecycle
+	Shutdowner hive.Shutdowner
+
+	ClientSet    k8sClient.Clientset
```
```suggestion
	Clientset    k8sClient.Clientset
```

Let's stick with the original form: https://pkg.go.dev/k8s.io/client-go/kubernetes#Clientset.

dylandreimerink committed:
operator: Add LB-IPAM

This commit adds the LB-IPAM feature. LB-IPAM allows users to specify
a set of pools containing one or more CIDRs. Services of type
LoadBalancer will receive Ingress IPs from these pools.

LB-IPAM is part of the ongoing work to add service announcements to
the BGP Control Plane. However, the component is designed to be generic
so it can be used by other features as well.

Co-authored-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

aanm commented on operator/pkg/lbipam/pool_store.go:
```diff
@@ -0,0 +1,36 @@
+func (ps *poolStore) Upsert(pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) {
+	if pool == nil {
+		return
+	}
```
Small nit that we are checking for `pool == nil` here but not on line 34. It would probably make sense to remove this check or added it on line 34.

aanm commented on operator/pkg/lbipam/pool_store.go:
```diff
@@ -0,0 +1,36 @@
+	return poolStore{
+		pools: make(map[string]*cilium_api_v2alpha1.CiliumLoadBalancerIPPool),
+	}
+}
```
What are we gaining with having a wrapper  for this map? The methods for this structure are not providing more value than the "normal" map operations.

dylandreimerink replied:
Yea, you are right. It once did, but it doesn't anymore, removing!

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are fully drained to avoid
+	// blocking the resources in case they are re-used after we shutdown.
```
IMO we shouldn't allow re-usage after shutting down.

dylandreimerink replied:
@joamaki suggested that I make it like this to ensure I leave the Resource[T] in a good state, I believe it has to do with no leaking goroutines that are blocking while testing ect. Perhaps the comment is a bit misleading in that case.

joamaki replied:
Yeah the point here is that we always need to consume the streams to the end, even if we cancel. If we don't, we may end up blocking them [1], which at the very least will prevent graceful termination if there are other observers for the same resource. Reword the comment to not talk about re-use, but just mention that this is done to drain in order to not block.

[1]: In case of `Resource[T]` we have a goroutine pulling from a work queue and calling the `next` function on each of the items. If we use `ToChannel` that `next` function will be sending to a channel. If we don't drain, then that function blocks on trying to send to the (by default unbuffered) channel. Thus any use of `ToChannel` should always be written to always drain the channel, regardless of whether or not we're observing over the whole lifetime of the program. Might perhaps make sense redesign `ToChannel` to make the draining part of the API. I'll think about this a bit.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+}
+
+func (ipam *LBIPAM) init(
+	ctx context.Context,
```
nit: this ctx is not being used anywhere in the function.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+	// Retry any pool related errors before we start processing the services. If we start processing services while we
+	// are missing pools due to connectivity issues, we might remove ingress IPs and cause service disruption.
+
+	const retryTimeout = 5 * time.Second
+	const retryCount = 6
```
Why 30 seconds in total?

dylandreimerink replied:
It was arbitrary, I added a comment explaining as much.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+
+func (ipam *LBIPAM) poolOnUpsert(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Deep copy so we get a version we are allowed to update
+	pool = pool.DeepCopy()
```
From what I can see we are not modifying the `pool` object so we don't need to perform DeepCopy.

dylandreimerink replied:
We modify the pool's status, since we don't store the pool status seperately we have to deep copy the whole thing so we can take ownership and don't have to worry about changing a shared copy owned by the store.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+
+func (ipam *LBIPAM) svcOnUpsert(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
```
From what I can see we are not modifying the `svc` object so we don't need to perform DeepCopy.

dylandreimerink replied:
We are already deep copying the status so this full copy is indeed not necessary, removed!

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+
+func (ipam *LBIPAM) svcOnDelete(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
```
From what I can see we are not modifying the `svc` object so we don't need to perform DeepCopy.

dylandreimerink replied:
We are already deep copying the status so this full copy is indeed not necessary, removed!

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+// on if the service requests are satisfied or not.
+func (ipam *LBIPAM) handleUpsertService(svc *slim_core_v1.Service) error {
+	key := resource.NewKey(svc)
+	sv, found, _ := ipam.serviceStore.GetService(key)
+	if !found {
+		sv = &ServiceView{
+			Key: key,
+		}
+	}
```
Here would be the place where we deepcopy the `sv` if we found it in the `ipam.serviceStore`

dylandreimerink replied:
`sv` is a ServiceView object, which is owned by the LB-IPAM controller, no external references exist, so there is no need to perform a deep copy here.

aanm replied:
I thought this was a k8s store.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+	// We are responsible for this service.
+
+	// Update the service view
+	sv.Generation = svc.Generation
+	sv.Labels = svcLabels(svc)
+	sv.RequestedFamilies.IPv4, sv.RequestedFamilies.IPv6 = ipam.serviceIPFamilyRequest(svc)
+	sv.RequestedIPs = getSVCRequestedIPs(svc)
+	sv.Status = svc.Status.DeepCopy()
```
Because we are modifying `sv` here

dylandreimerink replied:
Modifying `sv` isn't an issue.

aanm commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1797 @@
+	// We are responsible for this service.
+
+	// Update the service view
+	sv.Generation = svc.Generation
+	sv.Labels = svcLabels(svc)
+	sv.RequestedFamilies.IPv4, sv.RequestedFamilies.IPv6 = ipam.serviceIPFamilyRequest(svc)
+	sv.RequestedIPs = getSVCRequestedIPs(svc)
+	sv.Status = svc.Status.DeepCopy()
```
We don't need to perform this `svc.Status.DeepCopy()`

dylandreimerink replied:
We need to deep copy the status here since the object and its status are owned by the store. We do modify this status object so we can later patch it, this would cause the shared object in the store to change, which isn't allowed to avoid strange bugs with other components that also use the status.

oblazek commented on Documentation/network/lb-ipam.rst:
```diff
@@ -0,0 +1,402 @@
+********************************************
+
+LB IPAM is a feature that allows Cilium to assign IP addresses to Services of
+type ``LoadBalancer``. This functionally is usually left up to a cloud provider,
```
```suggestion
type ``LoadBalancer``. This functionality is usually left up to a cloud provider,
```
?

ldelossa commented on pkg/k8s/apis/cilium.io/v2alpha1/types.go:
```diff
@@ -46,6 +46,9 @@ type CiliumEgressNATPolicyList struct {
 // +kubebuilder:validation:Pattern=`^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\/([0-9]|[1-2][0-9]|3[0-2])$`
 type IPv4CIDR string
 
+// +kubebuilder:validation:Pattern=`(^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25 ...
```
Can we avoid the crazy regexes with "format=cidr". See: https://github.com/cilium/cilium/blob/40fc0909fab479cf08f37088c4e09e2ef2bce725/pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go#L73

Same as above. 

You'll have to do ipv4 vs ipv6 validation in code, but using the cidr format is nice.

dylandreimerink replied:
Didn't know this was a think, I thought format was only a hint, not actually enforced. Thx.

ldelossa commented on clustermesh-apiserver/main.go:
```diff
@@ -100,10 +102,32 @@ var (
+	"Resources",
+
+	cell.Provide(
+		func(lc hive.Lifecycle, c k8sClient.Clientset) (resource.Resource[*slim_corev1.Service], error) {
```
nit: probably a bit cleaner to define this in a var above and not inline, its beginning to nest quite a bit.

ldelossa commented on operator/cmd/resources.go:
```diff
@@ -0,0 +1,61 @@
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_core_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
```
nit: better to alias this to something more specific "k8sutils"

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+	DaemonConfig *option.DaemonConfig
+}
+
+func newLBIPAM(params LBIPAMParams) *LBIPAM {
```
Should this constructor take a "ctx", and then you derive a life cycle one from it? This would allow the constructor (which I assume is the daemon here?) to stop the LB-IPAM subsys.

dylandreimerink replied:
LB-IPAM as built as a hive cell, life cycle + ctx works a little different with hive.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+	ipamCtx, cancel := context.WithCancel(context.Background())
+
+	lbIPAM := &LBIPAM{
+		ctx:          ipamCtx,
```
Its not a great practice to hold on a ctx or cancel func. Is this necessary? 

Typically, the ctx will flow to an event loop of sorts, and be checked in the loop (one example). 

Consider whether you really need to hold on to CTX and cancel function.

dylandreimerink replied:
Perhaps we need to discuss this with @joamaki, I moved the context to a field as part of his PR feedback. Before I did pass it via parameters exclusively but it is only really used when patching, so we ended up with a lot of params which were just used to pass down. This safes a lot of code that way.

In any case, we need to store the cancel since the Stop function needs access to it.

I agree with the idiom of passing ctx as the first argument when talking about API's I am personally not convinced in this case since it is always an internal object.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
```
```suggestion
// LBIPAM is the loadbalancer IP address manager and watcher/controller which allocates and assigns IP addresses
```

Since you're saying watcher/controller is an analogy for each other.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+
+	ipam.wg.Add(1)
+
+	// Start go routine, we can't block in OnStart
```
Either you mean "in Start" here, or this comment is missing context. What is "OnStart" ? If its not visible here, it needs to be removed or further explained.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+	ipam.cancel()
+
+	// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+	wgDone := make(chan struct{})
```
The need to do this appears a bit odd to me. 

Could you actually get away with using a channel everywhere instead of the wait group? It appears you always use a wait group of size 1? If that's the case, won't a plain channel accomplish a similar goal, and then you don't need to "convert" it.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+
+// preInit blocks until a pool object exists in the store
+func (ipam *LBIPAM) preInit() error {
+	preInitCtx, cancel := context.WithCancel(ipam.ctx)
```
hmm, the ctx usage here is odd. 

Really, the constructor of LB-IPAM should provide a context to the "Start" method, and then it should be passed thru the call stack. 

For instance, preInit should receive a "ctx" and if then if it needs some timeout behavior internally, it can derive a context from it. 

We probably should not be holding onto `ipam.ctx` and instead thread a context thru from the Start() method and on.

dylandreimerink replied:
This ctx is here to manage the lifetime of `stream.ToChannel` so we can signal it to stop. It is a sub-ctx of `ipam.ctx` to that it is canceled when the outer ctx is canceled.

We do not want to timeout here, we want to intentionally block and not start the full control until we see at least 1 Pool which allows us to always enable the feature (no flag) without much overhead.

When we do see a pool, we cancel this `ToChannel` and continue on the the actual init process. I believe this is a very valid usage of a context.

joamaki replied:
The context usage is fine here with `ToChannel`. I would however consider removing the whole `preInit` and just subscribing to pools only once (e.g. in Run() line 214) and then waiting for the first to show up before subscribing to services. You might have little bit of duplication with handling the first event, but it should still be less code overall.

joamaki replied:
You could even do this from within the `for select` loop and swap `svcChan` into a real one from an empty one. 

EDIT: You'll need to redo `init` so that it'll first subscribe to pools, sync them and block waiting for the first real event and not just sync, and then handle the first event and then subscribe to services and handle the rest.

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+	}
+	poolRetryQueue := workqueue.New()
+
+	// First sync all pools
```
nit, refactor this out to its own function?

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+		}
+	}
+
+	// Then sync all services
```
nit, refactor this out to its own function?

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+}
+
+// unmarkPool removes the "Conflicting" status from the pool and removes the internally disabled flag from its ranges
+func (ipam *LBIPAM) unmarkPool(targetPool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
```
Would be cool if this became a method on `targetPool` instead.

dylandreimerink replied:
I think we can refactor this at some point, but doing so now would drastically delay the PR yet further, so I would like to postpone that to a separate PR.

ldelossa replied:
SGTM

ldelossa commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1791 @@
+}
+
+// markPoolConflicting marks the targetPool as "Conflicting" in its status and disables all of its ranges internally.
+func (ipam *LBIPAM) markPoolConflicting(
```
nit: as a general comment, it looks like LBIPAM has a large method set, and some of those methods could actually become part of the objects they receive as parameters. 

if it makes sense, consider inverting the responsibility to the structures themselves, not LBIPAM manipulating them.

dylandreimerink replied:
This is inline with feedback from @joamaki as well. Writing the logic this way is easy since all of the methods have all of the data available they might need.

In a future PR, we are going to attempt to split up this logic into separate cells so each part can request the objects they need and let hive sort out the dependencies and init order.

ldelossa reviewed (changes requested):
Have some concerns about how ctx is being utilized. Rest are nits.

dylandreimerink committed:
docs: Add LB-IPAM documentation

This commit adds documentation for the LB-IPAM feature.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1807 @@
+		ipam.logger.Debugf("Updating pool '%s'", pool.GetName())
+		err = ipam.handlePoolModified(pool)
+		if err != nil {
+			ipam.logger.WithError(err).Error("Error in handlePoolModified()")
```
Why is this and the error below ignored?

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1807 @@
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
```
Why are we logging the errors everywhere? Why not log once at the top? Don't we end up with many duplicate log lines for the same error?

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -0,0 +1,1807 @@
+}
+
+func (ipam *LBIPAM) svcOnUpsert(k resource.Key, svc *slim_core_v1.Service) error {
+	ipam.logger.Debugf("Added/updated service '%s/%s'", svc.GetNamespace(), svc.GetName())
```
`Debugf` is quite expensive even if it doesn't end up being printed. Consider removing these from the final version and add metrics instead.
